data_steward/cdr_cleaner/cleaning_rules/replace_standard_id_in_domain_tables.py

```diff
@@ -98,33 +103,37 @@
     dc.standard_concept IS NULL or dc.standard_concept = 'C'
 """)
 
+DUPLICATE_ID_SUB_QUERY = JINJA_ENV.from_string("""
+SELECT
+    a.src_id,
+    a.domain_table,
+    a.new_concept_id,
+    ROW_NUMBER() OVER(ORDER BY a.src_id, a.new_concept_id) + src.max_id AS dest_id
+FROM
+    `{{project}}.{{sandbox_dataset}}.{{logging_table}}` AS a
+JOIN (
+    SELECT src_id
+    FROM `{{project}}.{{sandbox_dataset}}.{{logging_table}}`
+    WHERE domain_table = '{{table_name}}'
+    GROUP BY src_id
+    HAVING COUNT(*) > 1 ) b
+ON 
+    a.src_id = b.src_id AND a.domain_table = '{{table_name}}'
+CROSS JOIN (
+    SELECT
+        MAX({{table_name}}_id) AS max_id
+    FROM `{{project}}.{{dataset}}.{{table_name}}` 
+) src
+""")
+
 DUPLICATE_ID_UPDATE_QUERY = JINJA_ENV.from_string("""
 UPDATE 
     `{{project}}.{{sandbox_dataset}}.{{logging_table}}` AS to_update 
 SET 
     to_update.dest_id = v.dest_id
 FROM 
 (
-    SELECT
-        a.src_id,
-        a.domain_table,
-        a.new_concept_id,
-        ROW_NUMBER() OVER(ORDER BY a.src_id, a.new_concept_id) + src.max_id AS dest_id
-    FROM
-        `{{project}}.{{sandbox_dataset}}.{{logging_table}}` AS a
-    JOIN (
-        SELECT src_id
-        FROM `{{project}}.{{sandbox_dataset}}.{{logging_table}}`
-        WHERE domain_table = '{{table_name}}'
-        GROUP BY src_id
-        HAVING COUNT(*) > 1 ) b
-    ON 
-        a.src_id = b.src_id AND a.domain_table = '{{table_name}}'
-    CROSS JOIN (
-        SELECT
-            MAX({{table_name}}_id) AS max_id
-        FROM `{{project}}.{{dataset}}.{{table_name}}` 
-    ) src
+    {{unioned_query}}
```

Comment:
I'm having trouble seeing what the full query ultimately ends up looking like. Have you considered using a single jinja template like this rather than filling out multiple template fragments and stitching them together?
```suggestion
    {%- for table_name in table_names %}
    SELECT
      a.src_id,
      a.domain_table,
      a.new_concept_id,
      ROW_NUMBER() OVER(ORDER BY a.src_id, a.new_concept_id) + src.max_id AS dest_id
    FROM
      `{{project}}.{{sandbox_dataset}}.{{logging_table}}` AS a
    JOIN (
        SELECT src_id
        FROM `{{project}}.{{sandbox_dataset}}.{{logging_table}}`
        WHERE domain_table = '{{table_name}}'
        GROUP BY src_id
        HAVING COUNT(*) > 1 ) b
    ON a.src_id = b.src_id AND a.domain_table = '{{table_name}}'
    CROSS JOIN (
      SELECT
        MAX({{table_name}}_id) AS max_id
      FROM `{{project}}.{{dataset}}.{{table_name}}`
    ) src
    {% if loop.nextitem is defined %}
    UNION ALL
    {% endif -%}
    {% endfor %}
```
The query would then be rendered like this:
```python
DUPLICATE_ID_UPDATE_QUERY.render(
            project=self.project_id,
            sandbox_dataset=self.sandbox_dataset_id,
            logging_table=SRC_CONCEPT_ID_TABLE_NAME,
            table_names=self.affected_tables)
```
Because there are fewer variables and functions to track it is easier to follow the query generation logic. Less code also means less testing. It's also easier to visualize the full query when it's in one place. Selective use of `-` in jinja blocks enables fine control over whitespace when balancing lean vs pretty SQL.

Reply:
I have not considered it but this is certainly much cooler, I will keep that in mind for the future query generation.

Reply:
Okay thanks for this suggestion, I haven't used jinja much in the past now I feel like I have another option for quickly generating queries, very cool. In the beginning, I thought it would be tricky to write a single jinja template (SRC_CONCEPT_ID_MAPPING_QUERY) for generating the UNION query for `_logging_standard_concept_id_replacement` because for each table in the loop I had to call two helper functions to get the {domain}_concept_id and {domain}_source_concept_id, and these two variables were also passed into the jinja template. Then it hit me, functions are just objects like variables so I could just pass the helper functions to the jinja template and run it there. This is a good lesson for me, appreciate it.